### ISSUE TITLE
Why to compute value functions for half embedding vectors in q_module.py

### ISSUE THREAD

**Issue by changy12** [griver/Q-RAG#8]
In https://github.com/griver/Q-RAG/blob/pqn-qa/rl/q_module.py, I found value functions are computed for the first and second half embedding vectors, as shown in the following 3 parts. Why not compute one value for the whole embedding vectors? 

(1) In class TextQNet, why

D = s_embed.shape[-1] // 2
        logits_1 = (s_embed[:, :D] * a_embed[:, :D]).sum(-1) 
        logits_2 = (s_embed[:, D:] * a_embed[:, D:]).sum(-1) ?

Why not

return (s_embed * a_embed).sum(-1) ?

(2) In class TextVNext, why

        D = s_embed.shape[-1] // 2
        logits_1 = (s_embed[:, :, :D] * a_embeds[:, :, :D]).sum(-1) 
        logits_2 = (s_embed[:, :, D:] * a_embeds[:, :, D:]).sum(-1) 

        top_k_actions = min(logits_1.size(1), self.top_k_actions)
        top_ids_1 = torch.topk(logits_1, top_k_actions, dim=1).indices
        top_mask_1 = torch.zeros_like(logits_1 > 0).scatter_(1, top_ids_1, True)

        top_ids_2 = torch.topk(logits_2, top_k_actions, dim=1).indices
        top_mask_2 = torch.zeros_like(logits_2 > 0).scatter_(1, top_ids_2, True)

        v1 = alpha * logsumexp(logits_1 / alpha, attention_mask=s.available_mask & top_mask_1, dim=-1)
        v2 = alpha * logsumexp(logits_2 / alpha, attention_mask=s.available_mask & top_mask_2, dim=-1)?
    
Why not

       logits = (s_embed * a_embeds).sum(-1) 
       top_k_actions = min(logits.size(1), self.top_k_actions)
       top_ids = torch.topk(logits, top_k_actions, dim=1).indices
       top_mask = torch.zeros_like(logits > 0).scatter_(1, top_ids, True)
       v = alpha * logsumexp(logits / alpha, attention_mask=s.available_mask & top_mask, dim=-1)?

(3) In class TextMaxQNet, why 
            D = s_embed.shape[-1] // 2
            logits_1 = (s_embed[:, :, :D] * a_embeds[:, :, :D]).sum(-1) 
            logits_2 = (s_embed[:, :, D:] * a_embeds[:, :, D:]).sum(-1) 

            logits_1[s.available_mask == False] = torch.min(logits_1)
            logits_2[s.available_mask == False] = torch.min(logits_2)

            return (logits_1.max(-1).values, 
                    logits_2.max(-1).values)?

Why not

            logits = (s_embed * a_embeds).sum(-1)
            logits[s.available_mask == False] = torch.min(logits)
            return (logits.max(-1).values,0)?

Also (4) in ~/rl/agents/pqn.py
Why 
qf_loss = 0.5 * F.mse_loss(2 * qf_1, reward_batch) + 0.5 * F.mse_loss(2 * qf_2, reward_batch) ?

Why not
qf_loss = F.mse_loss(qf_1 + qf_2, reward_batch)?

This looks weird since we want q=qf_1+qf_2 to be as close to reward_batch (G_t^k) as possible as shown in Eq. (5) of the ICLR26 Q-RAG paper https://openreview.net/pdf?id=MS9nWFY7LG . In the former code, it seems to encourage 2 * qf_1 ~ 2 * qf_2 ~ reward_batch (~ means approximate equal). 

Thanks.

**Comment by nazarblch** [comment]
Thanks for the close reading of the code.

You correctly note that Q_1 + Q_2 ≈ G_t. The two‑headed value estimation can be viewed as a simplified variant of the double estimator idea. In classic Double Q‑learning, you maintain two separate Q‑networks to decorrelate action selection from evaluation, thereby reducing overestimation bias. Here, instead of two full networks, we split the representation itself and compute two values (qf_1 and qf_2) from a shared backbone. This provides similar training stability benefits. It works better in our experiments than F.mse_loss(qf_1 + qf_2, reward_batch).

This simplified form of the double estimator is not without limitations. If you have thoughts on how to improve this mechanism it would be interesting to discuss.

**Comment by changy12** [comment]
Thanks @nazarblch.

In the original double Q learning paper [1], the two Q functions both approximate the original Q function, i.e., Q_1 ≈ Q_2 ≈ G_t instead of G_t≈Q_1+Q_2. This I think is a major difference from your code. 

[1] https://proceedings.neurips.cc/paper_files/paper/2010/file/091d584fced301b442654dd8c23b3fc9-Paper.pdf 

Which double Q learning method/work did you refer to?

**Comment by nazarblch** [comment]
Both formulations are equivalent up to a constant factor.  
We train the components such that $2Q_1 \approx 2Q_2 \approx G_t$,  
which implies $Q_1 + Q_2 \approx G_t$.  

Consequently, each component approximates the same target (up to scaling),  
consistent with the principle of Double Q‑learning [1].

**Comment by changy12** [comment]
Got it. 
Thanks.
